### PR TITLE
POC Reactive GetMessageList

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR] Errors: 
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraFilterTest>FilterTest.messageShouldBeAppendedInInboxWhenFromDoesNotMatchRule:1082 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraFilterTest>FilterTest.messageShouldBeAppendedInSpecificMailboxWhenFromRuleMatches:748 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraQuotaMailingTest>QuotaMailingTest.lambda$shouldSendANoticeWhenThresholdExceeded$0:112 » JsonPath
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.attachmentsShouldBeRetrievedWhenChainingSetMessagesAndGetMessagesTextAttachment:4371 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.sendingAMailShouldLeadToAppropriateMailboxCountersOnOutbox:1079 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.sendingAMailShouldLeadToAppropriateMailboxCountersOnSent:1124 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldDeliverMessageToRecipient:3044 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldMoveMessageInSentWhenMessageIsSent:2527 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldSendMessageByMovingDraftToOutbox:2258 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldUpdateIsAnsweredWhenInReplyToHeaderSentViaOutbox:4888->SetMessagesMethodTest.receiveFirstMessage:5138 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.setMessagesShouldUpdateIsForwardedWhenXForwardedHeaderSentViaOutbox:4935->SetMessagesMethodTest.receiveFirstMessage:5138 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [ERROR]   CassandraVacationIntegrationTest>VacationIntegrationTest.jmapVacationShouldGenerateAReplyWhenActive:130 » ConditionTimeout
[69b3b13dcd613aa1a0db780c25f3a7c64c4ac1af] [INFO] 
```

**Comment by chibenwa** [comment]
test this please

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java [hunk 1]
```diff
@@ -251,7 +253,7 @@
      * @param session
      *            the context for this call, not null
      */
-    List<MessageId> search(MultimailboxesSearchQuery expression, MailboxSession session, long limit) throws MailboxException;
+    Flux<MessageId> search(MultimailboxesSearchQuery expression, MailboxSession session, long limit) throws MailboxException;
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java:256]: should be a `Publisher`? (and maybe also in other APIs)

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java [hunk 2]
```diff
@@ -132,19 +132,15 @@ private MDCBuilder mdc(GetMessageListRequest messageListRequest) {
     }
 
     private Mono<GetMessageListResponse> getMessageListResponse(GetMessageListRequest messageListRequest, MailboxSession mailboxSession) {
-        try {
-            MultimailboxesSearchQuery searchQuery = convertToSearchQuery(messageListRequest);
-            Long postionValue = messageListRequest.getPosition().map(Number::asLong).orElse(DEFAULT_POSITION);
+        MultimailboxesSearchQuery searchQuery = convertToSearchQuery(messageListRequest);
+        Long postionValue = messageListRequest.getPosition().map(Number::asLong).orElse(DEFAULT_POSITION);
```
Comment (Arsnael) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java:136]: s/postionValue/positionValue

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 3]
```diff
@@ -677,26 +676,23 @@ private boolean hasChildIn(Mailbox parentMailbox, List<Mailbox> mailboxesWithPat
...
-        ImmutableSet<MailboxId> wantedMailboxesId =
-            getInMailboxes(expression.getInMailboxes(), session)
-                .filter(id -> !expression.getNotInMailboxes().contains(id))
-                .collect(Guavate.toImmutableSet());
-
-        return index.search(session, wantedMailboxesId, expression.getSearchQuery(), limit);
+    public Flux<MessageId> search(MultimailboxesSearchQuery expression, MailboxSession session, long limit) throws MailboxException {
+        return getInMailboxes(expression.getInMailboxes(), session)
+            .filter(id -> !expression.getNotInMailboxes().contains(id))
+            .collect(Guavate.toImmutableSet())
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:682]: what about using windowing?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java [hunk 4]
```diff
@@ -105,38 +118,29 @@
...
-                postionValue + messageListRequest.getLimit().map(Number::asLong).orElse(maximumLimit))
-                .stream()
-                .skip(postionValue)
-                .forEach(builder::messageId);
-            return builder.build();
-        } catch (MailboxException e) {
-            throw new RuntimeException(e);
-        }
+    private Mono<GetMessageListResponse> getMessageListResponse(GetMessageListRequest messageListRequest, MailboxSession mailboxSession) {
+        MultimailboxesSearchQuery searchQuery = convertToSearchQuery(messageListRequest);
```
Comment (mbaechler) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java:135]: you should wrap than in a Mono (and run it on `parallel` Scheduler)

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java [hunk 5]
```diff
@@ -105,38 +118,29 @@
...
+        Long positionValue = messageListRequest.getPosition().map(Number::asLong).orElse(DEFAULT_POSITION);
+        long limit = positionValue + messageListRequest.getLimit().map(Number::asLong).orElse(maximumLimit);
+
+        return Mono.defer(Throwing.supplier(() ->
```
Comment (mbaechler) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java:139]: why do you need to `defer`? search is not returning a Publisher?

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java [hunk 6]
```diff
@@ -105,38 +118,29 @@
...
+            Flux.from( mailboxManager.search(searchQuery, mailboxSession, limit))
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/GetMessageListMethod.java:140]: extra space

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:685]: Not relevant as we rely on ElasticSearch to filter mailboxes all at one

**Comment by Arsnael** [comment]
CI restart

test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 3 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:681]: StoreMailboxManager has nothing to do with ES.
If we expect to always have all MessageIds at once, we should declare a `Mono<List<MessageId>>`

**Comment by chibenwa** [comment]
## Before

Capture d’écran de 2020-04-10 13-22-11

## After

Capture d’écran de 2020-04-10 13-28-56

## Conclusion

Overall stable mean time, p99 decrease for getMailboxes (-50%) and getMessageList (-23%)

**Comment by mbaechler** [comment]
Did you inverted the results? I read the exact opposite of what you state

**Comment by chibenwa** [comment]
> Did you inverted the results? I read the exact opposite of what you state

No I did not

Again Max query time is not a relevant metric to me.

We'd need several runs to make it meaningful statistically.

**Comment by chibenwa** [comment]
Anyway, I have a rebase confict with the optimize branch thus need to rebase & re-test.

**Comment by chibenwa** [comment]
```

[64078072cdcf4940eb4804419436118d42ffc1ae] [ERROR] Failures: 
[64078072cdcf4940eb4804419436118d42ffc1ae] [ERROR]   CassandraMessageIdMapperTest>MessageIdMapperTest.copyInMailboxShouldThrowWhenMailboxDoesntExist:216 
[64078072cdcf4940eb4804419436118d42ffc1ae] Expecting code to raise a throwable.
[64078072cdcf4940eb4804419436118d42ffc1ae] [ERROR]   CassandraMessageIdMapperTest>MessageIdMapperTest.saveShouldThrowWhenMailboxDoesntExist:171 
```

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:681]: Well the usual pattern with a search index is "push all of your constraint", let it deal with ordering/sorting, and consume it's results.

Some windowing on the mailboxIds is a non sens, independently from ES

**Comment by chibenwa** [comment]
## Before

Capture d’écran de 2020-04-14 10-05-36


## After

Capture d’écran de 2020-04-14 10-05-41

## Conclusion

For GetMessageList specifically:
 - Stable mean and p50 latencies
 - Improvment of p99 (-16%)

**Comment by chibenwa** [comment]
test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 3 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:681]: ok

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**